core/src/engine/single_waiter_mutex.cpp

```diff
@@ -0,0 +1,29 @@
+#include <userver/engine/single_waiter_mutex.hpp>
+
+#include <engine/impl/mutex_impl.hpp>
+
+USERVER_NAMESPACE_BEGIN
+
+namespace engine {
+
+struct SingleWaiterMutex::Impl final {
+  impl::MutexImpl<impl::WaitListLight> mutex;
+};
+
+SingleWaiterMutex::SingleWaiterMutex() = default;
+
+SingleWaiterMutex::~SingleWaiterMutex() = default;
+
+void SingleWaiterMutex::lock() { impl_->mutex.lock(); }
```

Comment:
A sort of assertion "only 2 coroutines work with the mutex, no more" is needed here (in debug only) to make sure the mutex is not misused.

Reply:
Such checks have been made in the class WaitListLight.  In this place, it will be overcome easily in the case of thread racing.

https://github.com/userver-framework/userver/blob/d9a2218fa18cd37c9bcff2fb876323ebd49e7f09/core/src/engine/impl/wait_list_light.cpp#L82